pkg/artifact/downloader.go

```diff
@@ -0,0 +1,82 @@
+package artifact
+
+import (
+	"io"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path/filepath"
+
+	"github.com/square/p2/pkg/auth"
+	"github.com/square/p2/pkg/gzip"
+	"github.com/square/p2/pkg/uri"
+	"github.com/square/p2/pkg/util"
+)
+
+// Interface for downloading a single artifact.
+type Downloader interface {
+	// Downloads the artifact represented by the Downloader to the
+	// specified path and transfers file ownership to the specified user
+	DownloadTo(destination string, owner string) error
+}
+
+// Implements the Downloader interface. Simply fetches a .tar.gz file from a
+// configured URL and extracts it to the location passed to DownloadTo
+type locationDownloader struct {
+	location *url.URL
+	fetcher  uri.Fetcher
+	verifier auth.ArtifactVerifier
+}
+
+func NewLocationDownloader(location *url.URL, fetcher uri.Fetcher, verifier auth.ArtifactVerifier) Downloader {
+	return &locationDownloader{
+		location: location,
+		fetcher:  fetcher,
+		verifier: verifier,
+	}
+}
+
+func (l *locationDownloader) DownloadTo(dst string, owner string) error {
+	// Write to a temporary file for easy cleanup if the network transfer fails
+	// TODO: the end of the artifact URL may not always be suitable as a directory
+	// name
+	artifactFile, err := ioutil.TempFile("", filepath.Base(l.location.Path))
+	if err != nil {
+		return err
+	}
+	defer os.Remove(artifactFile.Name())
+	defer artifactFile.Close()
+
+	remoteData, err := l.fetcher.Open(l.location)
+	if err != nil {
+		return err
+	}
+	defer remoteData.Close()
+	_, err = io.Copy(artifactFile, remoteData)
+	if err != nil {
+		return util.Errorf("Could not copy artifact locally: %v", err)
+	}
+	// rewind once so we can ask the verifier
+	_, err = artifactFile.Seek(0, os.SEEK_SET)
+	if err != nil {
+		return util.Errorf("Could not reset artifact file position for verification: %v", err)
+	}
+
+	err = l.verifier.VerifyHoistArtifact(artifactFile, l.location)
```

Comment:
we OK with embedding this in? One could argue that now the downloader is doing more than just downloading. On the other hand, I don't have a good idea of what the alternative looks like (Separate downloader and verifier, then you download then verify?), so I don't yet have a better to suggest